apm-sniffer/apm-sdk-plugin/solrj-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/solrj/SolrClientInterceptor.java

```diff
@@ -0,0 +1,185 @@
+
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceConstructorInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.apache.skywalking.apm.plugin.solrj.commons.SolrjInstance;
+import org.apache.skywalking.apm.plugin.solrj.commons.SolrjTags;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.client.solrj.impl.HttpSolrClient;
+import org.apache.solr.client.solrj.request.AbstractUpdateRequest;
+import org.apache.solr.client.solrj.request.QueryRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.common.SolrDocumentList;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.params.UpdateParams;
+import org.apache.solr.common.util.NamedList;
+
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+
+public class SolrClientInterceptor implements InstanceMethodsAroundInterceptor, InstanceConstructorInterceptor {
+
+    @Override
+    public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+        SolrjInstance instance = new SolrjInstance();
+        HttpSolrClient client = (HttpSolrClient) objInst;
+
+        try {
+            URL url = new URL(client.getBaseURL());
+            instance.setRemotePeer(url.getHost() + ":" + url.getPort());
+
+            String path = url.getPath();
+            int idx = path.lastIndexOf('/');
+            if (idx > 0) {
+                instance.setCollection(path.substring(idx + 1));
+            }
+        } catch (MalformedURLException ignore) {
+        }
+        objInst.setSkyWalkingDynamicField(instance);
+    }
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        SolrRequest<?> request = (SolrRequest<?>) allArguments[0];
+        SolrjInstance instance = (SolrjInstance) objInst.getSkyWalkingDynamicField();
+
+        SolrParams params = getParams(request.getParams());
+        String collection = getCollection(instance, allArguments[2]);
+
+        AbstractSpan span = null;
+        if ("/update".equals(request.getPath())) {
```

Comment:
I talked with @IanCao , the key point is at here, you may not create a span, but you always close span. In that case, you will close an unexpected span. This causes a context leak after all.

Reply:
that means i need to verify the span is created by me or not. 
that is i have to check it on the same conditions.  @IanCao

Reply:
That verification will be very hard. I suggest dig deeper about why you don't need to create span in some case.

Or easy way, manipulate a parameter to propogate the creation status to after, in order to avoid recheck.

Reply:
i got it. thanks all of you.